4337/src/utils/execution.ts

```diff
@@ -79,8 +77,8 @@ export const safeApproveHash = async (
   if (!skipOnChainApproval) {
     if (!signer.provider) throw Error('Provider required for on-chain approval')
     const chainId = (await signer.provider.getNetwork()).chainId
-    const typedDataHash = utils.arrayify(calculateSafeTransactionHash(safe, safeTx, chainId))
-    const signerSafe = safe.connect(signer)
+    const typedDataHash = ethers.getBytes(await calculateSafeTransactionHash(safe, safeTx, chainId))
+    const signerSafe = safe.connect(signer) as typeof safe;
```

Comment:
I'm not sure if this is the idiomatic way of doing this. What is happening here is that `safe` is of type `Contract` (which behaves like a `Record<string, BaseContractMethod>`), but connect returns a `BaseContract` without access to methods.

The rationale for using `as typeof safe` instead of `as Contract` is to convey that `connect` returns the same thing as the original `safe` (regardless of what type it is), and would continue to work if we were to change `safe` to use a TypeChain generated strongly typed contract type.

Reply:
Hardhat toolbox has `typechain` types generation out of the box, so you could import the Safe type from the type folder:
```
import type { Safe } from '../../typechain-types/'
```

And then change the `safe` variable type to `Safe`, and then `.connect` should be correctly typed to return `Safe`. This may require adjustments to places where the function is used.

Reply:
One thing I noticed with `typechain` is that it adds a `connect` method with correct types, but it does not do the same for the `attach` method.

So, if you have:

```
let safe: Safe;
let otherSafe = safe.attach(otherAddress);
```

Then `otherSafe` will have a `BaseContract` type 😞.

Reply:
> One thing I noticed with `typechain` is that it adds a `connect` method with correct types, but it does not do the same for the `attach` method.
> 
> So, if you have:
> 
> ```ts
> let safe: Safe;
> let otherSafe = safe.attach(otherAddress);
> ```
> 
> Then `otherSafe` will have a `BaseContract` type 😞.

Yes, it is true, and that's annoying. But you can use the `getContractAt("ContractName", address)` (I hope that's the correct function name) and it'll be correctly typed.

Reply:
Oh, thanks.

It turns out that the TypeChain configuration was not setup correctly (missing including the type chain output directory in the tsconfig).

`getContractAt` worked like a charm - thanks for the tip.